Dev/chengwl/cmake build x86 compile

chengwl-qti opened the pull request:
Summary:
- Build libLiteRtCompilerPlugin_Qualcomm.so and realted Qualcomm test file on x86 
- Qualcomm build has been restructured from a single monolithic CMakeLists.txt into a modular

chunhsue-qti commented on litert/tools/CMakeLists.txt:
```diff
@@ -124,30 +148,39 @@ target_include_directories(run_model
+            ${ABSEIL_TOOL_LIBS}
+            absl::random_random
+            ${LITERT_TOOL_COMMON_LIBS}
+            tensorflow-lite
```
Can we move `tensorflow-lite` outside `"-Wl,--end-group"`?
Since there is already `LINK_GROUP:RESCAN` in `tensorflow-lite`.

graham0824 commented on litert/CMakeLists.txt:
```diff
@@ -345,7 +346,7 @@ if(NOT TARGET tensorflow-lite)
       IMPORTED_LOCATION "${TFLITE_BUILD_DIR}/libtensorflow-lite.a"
       INTERFACE_INCLUDE_DIRECTORIES "${TFLITE_SOURCE_DIR};${TFLITE_BUILD_DIR}"
-      INTERFACE_LINK_LIBRARIES "${_litert_tflite_link_libs}"
+      INTERFACE_LINK_LIBRARIES "$<LINK_GROUP:RESCAN,${_litert_tflite_link_libs}>"
```
Could we add a comment here to mention that "The LINK_GROUP:RESCAN" is aiming to resolve circular dependency issue?

graham0824 commented on litert/cc/CMakeLists.txt:
```diff
@@ -26,6 +26,7 @@ add_library(litert_cc_api STATIC
     litert_model.cc
     litert_opaque_options.cc
     litert_tensor_buffer.cc
+    internal/litert_extended_model.cc
```
Could we move this line to Line #24 to align the previous format?

chunhsue-qti replied:
The order matters in cmake, please check after moving it.

graham0824 commented on litert/tools/CMakeLists.txt:
```diff
@@ -124,30 +148,39 @@ target_include_directories(run_model
+        ${LITERT_TOOL_COMMON_LIBS}
+        tensorflow-lite
+        ${ABSEIL_TOOL_LIBS}
         absl::random_random
```
It seems the absl::random_random in not belonged to "ABSEIL_TOOL_LIBS", I'm thinking if we could remove Line #164 and then move absl::random_random in to the set "ABSEIL_TOOL_LIBS".

If the current code is aiming for reduce the size of apply_plugin_main (It appears that apply_plugin_main didn't need absl::random_random, and it links "ABSEIL_TOOL_LIBS"), then let's keep this code here.

chunhsue-qti commented on litert/vendors/qualcomm/compiler/CMakeLists.txt:
```diff
@@ -0,0 +1,62 @@
+        Threads::Threads
+        litert_c_api
+        litert_compiler_plugin
+        litert_runtime
+        litert_core
+        litert_core_model
+        litert_core_cache
+        litert_cc_api
+        litert_cc_options
+        litert_c_options
+        "-Wl,--end-group"
```
Can you check if we can use `litert_runtime_c_api_shared_lib` and get rid of directly depend on a lot of litert targets?

chunhsue-qti commented on litert/vendors/qualcomm/core/backends/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+)
+
+target_link_libraries(qnn_backends
+    PUBLIC
```
Is it possible to make it private?

chunhsue-qti commented on litert/vendors/qualcomm/core/CMakeLists.txt:
```diff
@@ -0,0 +1,64 @@
+add_library(qnn_core STATIC
+    # Root level
+    ../qnn_manager.cc
+    ../context_binary_info.cc
```
I think targets in the core folder should not include anything outside the folder. Can you please check?

chunhsue-qti commented on litert/vendors/qualcomm/dispatch/CMakeLists.txt:
```diff
@@ -0,0 +1,63 @@
+        "-Wl,-soname=libLiteRtDispatch_Qualcomm.so"
+    )
+
+    target_link_libraries(dispatch_api_qualcomm_so PRIVATE 
```
Can you please explain why we need this extra `target_link_libraries`?

chengwl-qti replied:
build on x86 need this target_link_libraries to solve undefined symbol

chunhsue-qti commented on litert/vendors/qualcomm/CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+# Fetch dependencies needed by core libraries (dump needs nlohmann_json)
+include(FetchContent)
+
+if(LITERT_OFFLINE_BUILD)
```
Let separate offline build in another PR.

chunhsue-qti commented on litert/vendors/qualcomm/core/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+target_link_libraries(qnn_core
+    PUBLIC
+        qnn_manager
+        qnn_context_binary_info
```
Can you check if we need these targets?

chunhsue-qti commented on litert/vendors/qualcomm/compiler/CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+        qnn_transformation
+        qnn_dump
+    PRIVATE
+        litert_cc_internal
```
Can you check why we need `litert_cc_internal`?

chunhsue-qti commented on litert/vendors/qualcomm/core/backends/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/..>
+        $<BUILD_INTERFACE:${TENSORFLOW_SOURCE_DIR}>
+        $<BUILD_INTERFACE:${TFLITE_SOURCE_DIR}>
```
```suggestion

```

chunhsue-qti commented on litert/vendors/qualcomm/core/backends/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+cmake_minimum_required(VERSION 3.20)
+
+# Collect all backend source files
+file(GLOB BACKEND_SRCS "*.cc")
```
Can we use `target_source` instead of global var like this for all targets.

chunhsue-qti commented on litert/vendors/qualcomm/core/CMakeLists.txt:
```diff
@@ -0,0 +1,59 @@
+
+# Add subdirectories for backends and wrappers
+add_subdirectory(backends)
+add_subdirectory(wrappers)
```
Place all `add_subdirectory` together.

chunhsue-qti commented on litert/vendors/qualcomm/core/CMakeLists.txt:
```diff
@@ -0,0 +1,59 @@
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/..>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+        $<BUILD_INTERFACE:${TENSORFLOW_SOURCE_DIR}>
+        $<BUILD_INTERFACE:${TFLITE_SOURCE_DIR}>
```
```suggestion

```

chunhsue-qti commented on litert/vendors/qualcomm/core/wrappers/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+        $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}>
+        $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/..>
+        $<BUILD_INTERFACE:${TENSORFLOW_SOURCE_DIR}>
+        $<BUILD_INTERFACE:${TFLITE_SOURCE_DIR}>
```
```suggestion
```

chunhsue-qti commented on litert/vendors/qualcomm/CMakeLists.txt:
```diff
@@ -0,0 +1,102 @@
+        $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}>
+        $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/..>
+        $<BUILD_INTERFACE:${TENSORFLOW_SOURCE_DIR}>
+        $<BUILD_INTERFACE:${TFLITE_SOURCE_DIR}>
```
```suggestion

```

chunhsue-qti commented on litert/vendors/qualcomm/CMakeLists.txt:
```diff
@@ -0,0 +1,102 @@
+        $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}>
+        $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/..>
+        $<BUILD_INTERFACE:${TENSORFLOW_SOURCE_DIR}>
+        $<BUILD_INTERFACE:${TFLITE_SOURCE_DIR}>
```
```suggestion
```

chunhsue-qti reviewed (changes requested):
Thanks for the great changes! Let's clean up a bit more!

ai-edge-bot committed:
Qualcomm AI Engine Direct - Build qualcomm compile so with cmake.

Summary:

- Build libLiteRtCompilerPlugin_Qualcomm.so and related Qualcomm test file on x86
- Qualcomm build has been restructured from a single monolithic CMakeLists.txt into a modular.
- Use "--start-group" instead of "--whole-archive" to reduce share.
  libraries size to 1MB.
